rfcs/20200929-keras-mixed-precision.md

```diff
@@ -0,0 +1,1152 @@
+
+### The global policy is not thread local
+
+The global policy will not be thread-local, although this may change in the future. The reason it is not thread local is that [`tf.distribute.MirroredStrategy`](https://www.tensorflow.org/api_docs/python/tf/distribute/MirroredStrategy) currently spawns threads, each which have their own versions of thread-local variables. MirroredStrategy will explicitly copy all the TensorFlow thread-local variables from the parent thread to its child threads, but since Keras is being moved outside TensorFlow, MirroredStrategy cannot depend on Keras and therefore it cannot copy the global policy.
+
+As a consequence of the global policy not being thread local, it is impossible to create a float32 model and a mixed precision in different threads in parallel.
+
+If in the future, MirroredStrategy stops using threads or exposes a mechanism to copy over thread-local variables to its child threads, we will consider changing the policy to be thread-local. Making the policy thread-local may break a user’s model if they spawn threads and create layers from within the threads, but it is unlikely users do this in practice.
+
+### Where should users set the global policy?
+
+It can be difficult to determine where a user should set the global policy to "mixed_float16" if they want to use mixed precision. For example, suppose they want to add mixed precision to a program that looks like this:
+
+```python
+def create_model(num_layers):
+  return tf.keras.Sequential([tf.keras.layers.Dense(10, activation="relu")
+                              for _ in range(num_layers)])
+
+def main(_):
+  model = create_model(num_layers=10)
+  model.compile(...)
+  dataset = ...
+  model.fit(dataset)
+
+if __name__ == "__main__":
+  app.run(main)
+```
+
+A `dtype` flag can be added which can be "float32" or "mixed_float16", but should the global dtype policy be set in `create_model` or `main`? In this case, we will recommend users set it in `main`. The issue is `create_model` appears to be stateless, so we should avoid setting state by calling `set_global_policy`. The recommend way to add mixed precision is therefore:
+
+```python
+flags.DEFINE_enum("dtype", "float32", ("float32", "mixed_float16"),
+                  description="Dtype policy to use")
+
+def create_model(num_layers):
+  return tf.keras.Sequential([tf.keras.layers.Dense(10, activation="relu")
+                              for _ in range(num_layers)])
+
+def main(_):
+  tf.keras.mixed_precision.set_global_policy(flags.FLAGS.dtype)
+  model = create_model(num_layers=10)
+  model.compile(...)
+  dataset = ...
+  model.fit(dataset)
+
+if __name__ == "__main__":
+  app.run(main)
+```
+
+Alternatively, the global policy can be set in `create_model`. This will work in cases where multiple models are not built in parallel using threads, and where every time a model is created, the global policy is set to an appropriate value.
```

Comment:
I am a little confused about the case here. This is the way for use case "where multiple models are **not** built in parallel using threads" or "where multiple models are built in parallel using threads"?

Reply:
I removed the part about threads and reworded this paragraph. I was trying to state that you cannot create a float32 and a mixed precision model in parallel, but I already state that in the preceding "The global policy is not thread local" section and that is true regardless of where the global policy is set.